designs/CB-OPT-design-week7-update.md

```diff
@@ -0,0 +1,224 @@
+y:int = 2
+a:int = 0
+a = 2 * (x+y) + 3 * (y + x) + 5 * (x+y)
+```
+
+**After Optimization**
+```python
+x:int = 1
+y:int = 2
+a:int = 0
+e:int = 0
+e = x + y
+a = 2 * e + 3 * e + 5 * e
+```
+
+### 10. Eliminate Redundant Code(Haven't Implemented yet)
+**Before Optimization**
+```python
+x:int = 1
+x = x
+```
+
+**After Optimization**
+```python
+x:int = 1
+```
+
+## B. Modification on AST and IR
+Both are not changed.
+
+## C. New Changes
+Two new files `optimize_ast.ts` and `optimize_ir.ts` are added to implement optimizations on `ast` and `ir` respectively. One debugger `debug_ir.ts` is added to print pretty `ir` representation and generate DOT graph. To use this debugger, a new dependency is added: `@diagrams-ts/graphviz-cli-renderer`. Also, the host machine should install [graphviz](https://graphviz.org/download/) as well. We also modified the funciton `run`s in `repl.ts` and `runner.ts` by adding two boolean parameters to indicate the optimization option. The optimization is disabled by default. In addition, to observe the length of the optimized WASM code for testing, we also added a new field `watCode` in class `BasicREPL` in order to fetch the generated wasm code after running.
+
+We have implemented a majority of our design decisions in "optimize_ast.ts". We have implemented dead code elimination, including unreachable instructions after `return` and dead branches such as `if True` and `while False`. Our program can also do constant folding for all operations whose operands are constant. Further folding will be available when the constant propagation is facilitated.
+
+The optimization takes place in a iterative fashion until the source code remains unchanged during a pass. Currently, passes available are:
+
+* AST-Level: 
+  1. dead code elimination
+  2. constant folding
+* IR-Level:
+  1. constant folding
+
+
+## D. New decisions
+We plan to implement the following optimizations in the next sprint: 1. Constant propagation 2. Liveness Analysis. 
+
+Most likely, we will try to use control-flow anlyasis metioned in class for our implementation on constant propagation which indicates that we will implement constant propagtaion in "optimize_ir.ts" first. 
+
+We also plan to implement liveness analysis at `ir` first since the control flow at this level is less complex, but it does not preclude our intention to introduce this analysis at `ast` since removing dead variable at the earlier stage may bring more benefits, and we can exploit more control-flow semantics despite its complexity.
+
+We already observe noticable difference in the optimality of enforcing the same optimization techniques at different levels. For constant folding, if we can merge operations at `ast`, less intermediate variable will be generated compared with applying folding at `ir` level. Although these redundant variables may be eliminated by further propagation and folding at `ir`, wipping the **seed** out at `ast` can definitely simplify the further optimization. 
```

Comment:
Yes, but optimizing at `ast` comes with a huge price, you are dealing with a much more complex environment in `ast` and should be super careful with your optimizations. I'd start with IR, do lots of testing on it, and then look for super simple optimizations that can be done on `ast`.

Reply:
Thanks for your advice, I think we are planning the same. We will start our implementation from `IR` and make sure our optimization is not introducing any hard conflicts with other groups.